File: frontend/src/ai/obs.js
/**
 * build observation — espelho de backend/ai/obs.py.
 *
 * estrutura canônica (341 floats):
 *   [0..335]  336 floats raycasts (48 raios × 7: distance + 6 one-hot)
 *   [336]     speed normalizado [0, 1]
 *   [337]     angularVel ∈ [-1, 1]
 *   [338]     canKick ∈ {0, 1}
 *   [339]     ball_visible_anywhere ∈ {0, 1}
 *   [340]     action_repeat_idx / 4 ∈ {0, 0.25, 0.5, 0.75}
 *
 * por raio (7 floats):
 *   [0]  distance normalizada ∈ [0, 1]  OU  -1.0 (sentinela "none")
 *   [1]  one_hot wall          ∈ {0, 1}
 *   [2]  one_hot ball          ∈ {0, 1}
 *   [3]  one_hot ally          ∈ {0, 1}
 *   [4]  one_hot enemy         ∈ {0, 1}
 *   [5]  one_hot goal_own      ∈ {0, 1}
 *   [6]  one_hot goal_enemy    ∈ {0, 1}
 *
 * sentinela canônica: distance = -1.0, one_hot = [0,0,0,0,0,0] → "nada detectado".
 * mapeamento: wall=0, ball=1, ally=2, enemy=3, goal_own=4, goal_enemy=5.
 *
 * depende de AISoccer.Physics.castRays(player, state) retornar 96 floats
 * [dist, type_code, dist, type_code, ...] (ver frontend/src/game/physics.js).
 */
(function () {
    "use strict";

    var OBS_SIZE = 341;
    var NUM_RAYS = 48;
    var RAY_FEATURE_SIZE = 7;
    var STATE_EXTRA_SIZE = 5;
    var STATE_EXTRA_BASE = NUM_RAYS * RAY_FEATURE_SIZE; // 336

    // mapeamento canônico tipo → índice no one-hot. replicar exato em Python.
    var TYPE_TO_ONEHOT_IDX = {
        wall: 0,
        ball: 1,
        ally: 2,
        enemy: 3,
        goal_own: 4,
        goal_enemy: 5,
    };

    // bate com backend/config.py.
    var MAX_SPEED_PLAYER = 3.5;

    /**
     * decodifica o type_code (do TYPE_MAP de physics.js) num nome canônico.
     * espelho exato de physics.js::codeToType.
     */
    function _codeToTypeName(code) {
        if (code < -0.5) return "none";
        if (code < 0.1) return "wall";
        if (code < 0.3) return "ball";
        if (code < 0.5) return "ally";
        if (code < 0.7) return "enemy";
        if (code < 0.9) return "goal_own";
        return "goal_enemy";
    }

    /**
     * constrói o vetor de observação de 341 floats para o player.
     *
     * @param {Object} player Player vivo do GameState.
     * @param {Object} state  GameState completo.
     * @param {number} actionRepeatIdx frame atual no ciclo de action repeat (0..3).
     * @returns {Float32Array} length 341.
     * @throws {RangeError} se actionRepeatIdx fora de [0, 3] ou não-inteiro.
     */
    function gatherObs(player, state, actionRepeatIdx) {
        if (actionRepeatIdx === undefined) actionRepeatIdx = 0;
        if (
            !Number.isInteger(actionRepeatIdx) ||
            actionRepeatIdx < 0 ||
            actionRepeatIdx > 3
        ) {
            throw new RangeError(
                "actionRepeatIdx deve ser inteiro em [0, 3], recebido: " +
                actionRepeatIdx
            );
        }

        var obs = new Float32Array(OBS_SIZE);
        var rayData = AISoccer.Physics.castRays(player, state);

        var ballVisible = false;

        for (var i = 0; i < NUM_RAYS; i++) {
            var dist = rayData[i * 2];
            var typeCode = rayData[i * 2 + 1];
            var typeName = _codeToTypeName(typeCode);
            var base = i * RAY_FEATURE_SIZE;

            if (typeName === "none") {
                obs[base] = -1.0;
            } else {
                obs[base] = dist;
                var onehotIdx = TYPE_TO_ONEHOT_IDX[typeName];
                obs[base + 1 + onehotIdx] = 1.0;
                if (typeName === "ball") ballVisible = true;
            }
        }

        var speed =
            Math.sqrt(player.vx * player.vx + player.vy * player.vy) /
            MAX_SPEED_PLAYER;
        obs[STATE_EXTRA_BASE + 0] = Math.max(0.0, Math.min(1.0, speed));
        var rot = player.rot || 0.0;
        obs[STATE_EXTRA_BASE + 1] = Math.max(-1.0, Math.min(1.0, rot));
        obs[STATE_EXTRA_BASE + 2] = player.can_kick ? 1.0 : 0.0;
        obs[STATE_EXTRA_BASE + 3] = ballVisible ? 1.0 : 0.0;
        obs[STATE_EXTRA_BASE + 4] = actionRepeatIdx / 4.0;

        return obs;
    }

    var api = {
        OBS_SIZE: OBS_SIZE,
        NUM_RAYS: NUM_RAYS,
        RAY_FEATURE_SIZE: RAY_FEATURE_SIZE,
        STATE_EXTRA_SIZE: STATE_EXTRA_SIZE,
        TYPE_TO_ONEHOT_IDX: TYPE_TO_ONEHOT_IDX,
        gatherObs: gatherObs,
    };

    if (typeof window !== "undefined") {
        window.AISoccer = window.AISoccer || {};
        window.AISoccer.Obs = api;
    }
    if (typeof module !== "undefined" && module.exports) {
        module.exports = api;
    }
})();

File: backend/ai/obs.py
"""vetor de observação 341-dim do Player.

traduz a saída ordinal de raycasting.cast_rays (96 floats: 48 raios ×
[distância normalizada, tipo ordinal]) em vetor com one-hot por tipo +
5 floats de estado extra:

    obs[  0 .. 335]  →  336 floats — 48 raios × 7 floats por raio
    obs[336 .. 340]  →    5 floats — speed, angularVel, canKick,
                                      ball_visible_anywhere, action_repeat

por raio (7 floats):
    [0]  distance normalizada ∈ [0, 1]   OU  -1.0 (sentinela "none")
    [1]  one_hot wall          ∈ {0, 1}
    [2]  one_hot ball          ∈ {0, 1}
    [3]  one_hot ally          ∈ {0, 1}
    [4]  one_hot enemy         ∈ {0, 1}
    [5]  one_hot goal_own      ∈ {0, 1}
    [6]  one_hot goal_enemy    ∈ {0, 1}

sentinela canônica de ausência:
    distance = -1.0  e  one_hot = (0, 0, 0, 0, 0, 0)   (none não tem bit).

state extra (índices 336..340):
    [336]  speed                  = ‖(vx, vy)‖ / MAX_SPEED_PLAYER ∈ [0, 1]
    [337]  angularVel             = player.rot                    ∈ [-1, 1]
    [338]  canKick                = 1.0 se player.can_kick, senão 0.0
    [339]  ball_visible_anywhere  = 1.0 se algum raio detectou bola
    [340]  action_repeat_idx      = idx / 4.0                     ∈ [0, 0.75]

contrato Python ↔ JS: cada decisão (sentinela, índices do one-hot,
normalização do action_repeat, dtype float32) é replicada bit-a-bit no frontend.
"""

from __future__ import annotations

import math
from typing import TYPE_CHECKING

import numpy as np

from backend.config import MAX_SPEED_PLAYER, NUM_RAYS, TYPE_MAP
from backend.physics.raycasting import cast_rays

if TYPE_CHECKING:
    from backend.physics.entities import GameState, Player


NUM_RAYS_FROM_CONFIG: int = 48
RAY_FEATURE_SIZE: int = 7
STATE_EXTRA_SIZE: int = 5
OBS_SIZE: int = NUM_RAYS_FROM_CONFIG * RAY_FEATURE_SIZE + STATE_EXTRA_SIZE  # 341

assert NUM_RAYS == NUM_RAYS_FROM_CONFIG, (
    f"obs.py assume {NUM_RAYS_FROM_CONFIG} raios; config.NUM_RAYS={NUM_RAYS}"
)

# ordem fixa, replicada em JS. "none" não tem bit (sinalizado por distance=-1.0).
TYPE_TO_ONEHOT_IDX: dict[str, int] = {
    "wall":       0,
    "ball":       1,
    "ally":       2,
    "enemy":      3,
    "goal_own":   4,
    "goal_enemy": 5,
}

_FLOAT_TO_TYPE_STR: dict[float, str] = {v: k for k, v in TYPE_MAP.items()}

_STATE_EXTRA_BASE: int = NUM_RAYS_FROM_CONFIG * RAY_FEATURE_SIZE  # 336


def gather_obs(
    player: Player,
    state: GameState,
    action_repeat_idx: int = 0,
) -> np.ndarray:
    """constrói o vetor de observação de 341 floats float32 para o player.

    obs é construído da perspectiva do player. action_repeat_idx é o índice
    do frame atual no ciclo de action repeat (0..3).
    """
    if not 0 <= action_repeat_idx <= 3:
        raise ValueError(
            f"action_repeat_idx fora do range [0, 3]: {action_repeat_idx!r}"
        )

    obs = np.zeros(OBS_SIZE, dtype=np.float32)
    raw = cast_rays(player, state)

    ball_visible = False
    for i in range(NUM_RAYS_FROM_CONFIG):
        norm_dist = raw[2 * i]
        type_float = raw[2 * i + 1]
        type_str = _FLOAT_TO_TYPE_STR[type_float]
        base = i * RAY_FEATURE_SIZE

        if type_str == "none":
            obs[base] = -1.0
        else:
            obs[base] = norm_dist
            obs[base + 1 + TYPE_TO_ONEHOT_IDX[type_str]] = 1.0
            if type_str == "ball":
                ball_visible = True

    speed_norm = math.sqrt(player.vx * player.vx + player.vy * player.vy) / MAX_SPEED_PLAYER
    obs[_STATE_EXTRA_BASE + 0] = max(0.0, min(1.0, speed_norm))
    obs[_STATE_EXTRA_BASE + 1] = max(-1.0, min(1.0, float(player.rot)))
    obs[_STATE_EXTRA_BASE + 2] = 1.0 if player.can_kick else 0.0
    obs[_STATE_EXTRA_BASE + 3] = 1.0 if ball_visible else 0.0
    obs[_STATE_EXTRA_BASE + 4] = action_repeat_idx / 4.0

    return obs


def decode_ray_features(obs_vector: np.ndarray, ray_idx: int) -> dict:
    """helper de debug: decodifica o raio ray_idx em dict legível."""
    base = ray_idx * RAY_FEATURE_SIZE
    distance = float(obs_vector[base])
    raw_onehot = [float(x) for x in obs_vector[base + 1:base + RAY_FEATURE_SIZE]]

    if distance == -1.0:
        return {"distance": None, "detected_type": None, "raw_onehot": raw_onehot}

    detected_type = None
    for type_str, idx in TYPE_TO_ONEHOT_IDX.items():
        if raw_onehot[idx] == 1.0:
            detected_type = type_str
            break
    return {"distance": distance, "detected_type": detected_type, "raw_onehot": raw_onehot}
